src/fastpath/fastpath.c

```diff
@@ -6,15 +6,313 @@
+            * adds extra scheduler logic. Normally, this is done after the sched_context donate
+            * but after tweaking it, I don't immediately see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+
+            if (sc->scRefillMax > 0 && !thread_state_get_tcbInReleaseQueue(dest->tcbState)) {
+                if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+                schedulable = true;
+            }
+
+            if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+                crossnode = true;
+            }
+
+            /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            tcb_queue_t queue;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+
+            queue = ep_ptr_get_queue(ep_ptr);
+            queue = tcbEPDequeue(dest, queue);
+            ep_ptr_set_queue(ep_ptr, queue);
+
+            if (!queue.head) {
+                endpoint_ptr_set_state(ep_ptr, EPState_Idle);
+            }
+
+            reply_t *reply = REPLY_PTR(thread_state_get_replyObject(dest->tcbState));
+            if (reply != NULL) {
+                reply_unlink(reply, dest);
+            }
+
+            /* Now we donate the SC. The checks required for this were
+             * already done before the point of no return. */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+            if (!dest->tcbSchedContext) {
+                /* Equivalent to schedContext_donate without migrateTCB() */
+                sc->scTcb = dest;
+                dest->tcbSchedContext = sc;
+            }
+            assert(dest->tcbSchedContext);
+
+#ifdef ENABLE_SMP_SUPPORT
+            /* The part of migrateTCB() that doesn't involve the slowpathed FPU save */
+            dest->tcbAffinity = sc->scCore;
+#endif
+
+            /* Left this in the same form as the slowpath. Not sure if optimal */
+            if (sc_sporadic(dest->tcbSchedContext)) {
+                assert(dest->tcbSchedContext != NODE_STATE(ksCurSC));
+                if (dest->tcbSchedContext != NODE_STATE(ksCurSC)) {
+                    refill_unblock_check(dest->tcbSchedContext);
+                }
+            }
+
+            /* If dest was already not schedulable prior to the budget check
+             * the slowpath doesn't seem to do anything special besides just not
+             * not scheduling the dest thread. */
+            if (schedulable) {
+                if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority || crossnode) {
+                    SCHED_ENQUEUE(dest);
```

Comment:
This operation sets a flag on SMP kernels to perform doMaskReschedule() before the kernel exits if the destination thread belongs to a different core. If waking a thread on another core is going to be part of the FP, then doMaskReschedule will also need to be added before restore_user_context() in both branches.

Reply:
@alwin-joshy, did you get a chance to look into this?

Reply:
https://github.com/alwin-joshy/seL4/commit/565e043164443b643136eb19e22e3f3ad2b38500

Reply:
I agree with [your comment](https://github.com/seL4/seL4/pull/793/files/645632bfa3b1a90ece1c379c8cd9392fb82c05f8#r980638076), I missed that the slowpath was taken for the case where the destination thread is the new highest priority on another core as I explained [here](https://github.com/seL4/seL4/pull/793/files/645632bfa3b1a90ece1c379c8cd9392fb82c05f8#r1029138484).

One thing that appears to be missing from the slowpath is a remoteQueueUpdate for the case where postpone() puts the dest thread into the front of the release queue on another core and so the timer deadline needs to be adjusted.  Is my understanding correct that postpone() won't be called in the fastpath, and the slowpath would be taken instead? If this is the case, then I think the doMaskReschedule() can indeed be left out.

I'd suggest not extending the scope of this PR to include signalling higher-prio threads on a different core, and leave it for a future change.

Reply:
Yep, I think that is right. AFAIK `postpone()` only happens in `schedContext_resume()` when the sc fails the `!(refill_ready(sc) && refill_sufficient(sc, 0)` check. In the fastpath, the failure of this check results in a slowpath transition.